### PR TITLE
feat(native): deferred deep linking through the store install (TASK-20772)

### PR DESCRIPTION
**Pull request by kushagrasarathe**
## Summary

Adds a **dev-only preview page** at `/dev/migration` that showcases every UI surface for the PWA → app-store migration, composed from the **real app components** (not lookalikes) so the states read like production.

Surfaces covered:
- **01 Download prompt** — single variant, deadline in copy; mobile shows the visitor's store CTA, desktop shows the QR inline.
- **02 Access-ending screen** — full-screen after cutover (mirrors `ForceIOSPWAInstall`).
- **03 Guest link pages** — claim / request / invite, built from `PeanutActionDetailsCard` + `ActionListCard` + the real "Continue with Peanut" button.
- **04 Get-the-app banner** — the real home `CarouselCTA`.
- **05 Landing hero** — the real `<Hero>` component, CTA swapped to "Download now".
- **06 Review prompt** / **07 Notifications prompt** — `ActionModal` pre-prompts.
- **08 Scan-to-download** — a QR with an App Store / Google Play toggle (a browser can't tell iOS from Android).

Device-aware throughout: **one primary CTA per device** (mobile → store button; desktop → QR).

## Risk

**None to production.** Dev-only route — `/dev` is `notFound()` on prod (`dev/layout.tsx`), the whole page is behind that gate. No production code touched; only a new file under `src/app/(mobile-ui)/dev/`.

## QA

```
./scripts/dev   # or: PORT=3055 pnpm dev in this worktree
open http://localhost:3055/dev/migration
```
Toggle **iOS / Android / Desktop** (top-right) to see the device-aware CTAs; open each modal; on Desktop the download CTAs show the dual-store QR.

## Screenshots

Dev-only preview route — best viewed live at `/dev/migration` (steps above). Every surface is rendered from the real components; toggling the platform pill drives the CTA variants.

## Design notes

- No "native app" wording anywhere — users don't know the term; copy is "Peanut is becoming an app" / "download from the App Store or Google Play".
- The dual-store QR toggle is deliberate; a single smart link (`peanut.me/app` redirecting by device) would remove the toggle — noted in a code comment as the lighter alternative, pending a product call.
- Components are composed inline in the dev page for review — **not yet extracted** into production components; that happens when the migration decisions lock.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added deferred deep-linking restoration with destination, language, invitation, and campaign details, including Android/iOS handoff support.
  * Added a developer page to debug and simulate deferred payloads, with safe non-native environment blocking.
  * Added Play Store URL constant for UI verification.
* **Bug Fixes**
  * Prevent restored destinations from overriding already-handled in-app navigation and avoid “teleporting” when restoration is late or fails.
  * Improved retry/consumption behavior and stricter destination validation; preserve URL query/hash when deriving destinations.
* **Tests**
  * Expanded Jest coverage for deferred payload, restoration timing, retry, clipboard gating, and edge cases.
* **Chores**
  * Added Capacitor Preferences support for persisting restored state.

### CODE REVIEW

**Comment by kushagrasarathe** [comment]
@coderabbitai review

File: src/app/dev/deferred/page.tsx [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+                {payload && (
+                    <button
+                        className="border border-n-1 px-2 py-1"
+                        onClick={async () => {
+                            await copyIOSHandoff(payload)
+                            setCopied(true)
+                        }}
+                    >
+                        {copied ? 'copied ✓' : 'copy ios hand-off to clipboard'}
+                    </button>
```
[the remark this answers is not in the document]
Comment (kushagrasarathe) [reply src/app/dev/deferred/page.tsx:117]: Fixed in 5231ab7d1 — building a new payload now resets the copied indicator.

**Comment by jjramirezn** [comment]
## 🤖 High-effort code review — 10 verified findings

Automated multi-agent review (4 finders + independent verification per finding; 22 candidates, 3 refuted). The happy path works, but the restore path breaks in several everyday situations.

### 🔴 Blockers

**1. Existing users get locked out of login (`src/utils/deferred-link.ts:247`)**
`applyDeferredPayload` writes the `inviteCode` cookie for **30 days**. The setup page's `skipInviteGate = !!userInviteCode` then routes every launch straight to the **Signup** step — past the Landing screen, which is the only screen with a Log In button (same regression class as PR #2346, this time via inviteCode instead of campaignTag). An existing user who taps a friend's invite link and installs the app can't log in on that device for up to 30 days, since `useZeroDev` only clears the cookie on successful signup.

**2. Unmapped dests navigate to routes that don't exist (`src/utils/deferred-link.ts:255`)**
`deepLinkToNativePath(rawDest) ?? rawDest` drops the must-map guard that `openDeepLink` enforces. If mapping returns `null` (e.g. a stray `%` making `decodeURIComponent` throw), the raw string passes the same-origin check and gets `router.push`ed verbatim — but in the native static export `/send/<user>` only exists as `/send?recipient=...`, so the fresh-install user's very first screen is a 404/blank route. Should be dropped like `openDeepLink` drops it.

**3. Transient Android referrer failure burns the one-shot flag (`src/utils/deferred-link.ts:213`)**
`CONSUMED_KEY` is set even when `readInstallReferrer()` resolved `null` from a transient failure (5s timeout, `SERVICE_UNAVAILABLE`, slow Play Services on first boot). Invite code / campaign / dest are then lost forever — silently — even though the Play referrer stays readable for ~90 days and a retry on next launch would be safe and prompt-free on Android. Suggest: only mark consumed on Android after a definitive read (or a definitive `FEATURE_NOT_SUPPORTED`-class failure).

**4. Late restore yanks users out of mid-onboarding (`src/hooks/useNativePlugins.ts:92`)**
The un-awaited restore fires `router.push(restored.dest)` whenever it resolves, guarded only by `anyDeepLinkNavigated` (which tracks deep links, not user navigation). The Android read can take up to 5s and the iOS paste prompt can sit unresolved indefinitely — a user already tapping through signup/passkey creation gets teleported to `/claim/...` mid-flow.

### 🟡 Should fix

**5. iOS re-prompts after kill-during-prompt (`src/utils/deferred-link.ts:213`)**
On iOS the consumed flag is only written after `Clipboard.read()` resolves. Backgrounding/killing the app while the system paste prompt is up (a common reflex) means the flag is never set — every subsequent cold start with a web URL on the clipboard re-raises the paste prompt, breaking the "never re-prompt" invariant and reading as repeated clipboard snooping.

**6. `cap sync` swept `@capacitor/camera` into the Android build (`android/capacitor.settings.gradle:11`, `android/app/capacitor.build.gradle:14`)**
The regenerated gradle files add the previously-absent camera native module — unrelated to this PR. The next Play binary (the one this PR needs shipped) gains camera/media merged-manifest permission entries, which can trip Play Console's sensitive-permission review or surface unexplained permissions to users. Either exclude it or ship it as a deliberate, declared change.

**7. First-launch race: invite gate vs. referrer read (`src/utils/deferred-link.ts:247`)**
Setup's `determineInitialStep` reads the `inviteCode` cookie ~100ms after mount, but the Android referrer read can take up to 5s before the cookie is written. On exactly the launch the feature was built for, the invite-driven "straight to signup" path degrades to the generic Landing step (and the later dest push may then yank the user — see #4).

**8. `stripLocalePrefix` misses locale-with-query (`src/utils/deferred-link.ts:84`)**
Splitting on `/` without separating the query string means `/pt-br?x=1` fails `isValidLocale` (first segment is `pt-br?x=1`) and the locale prefix survives into the native dest → 404 in the static export.

### 🧹 Cleanup

**9. Dev page forks the prod wall (`src/app/dev/deferred/page.tsx:65`)** — re-implements `(mobile-ui)/dev/layout.tsx`'s prod block inline with a native exception instead of generalizing the shared `DevLayout` (e.g. a native-allowed flag) and living under `(mobile-ui)/dev/` like its siblings. Future policy changes to the wall won't apply here.

**10. Inline invite normalization duplicates `toInviteCode` (`src/utils/deferred-link.ts:246`)** — `payload.invite?.trim().replace(/^@/, '').toLowerCase()` is byte-for-byte `toInviteCode` from `general.utils` (already imported from). Call the helper so the paths can't drift.

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by innolope-dev** [comment]
@kushagrasarathe — I'd explored this problem in a design plan a couple of days before this PR went up, and landed on the same two channels you built (Play Install Referrer + iOS clipboard). Having compared the two, **your serverless shape is the better call**, and I want to be specific about why, since my version had a server tier and I now think that was wrong:

- Mine added two public unauthenticated endpoints plus a TTL store to the API to carry locale, invite code, campaign tag and dest. Invite codes are already public (derived from usernames, resolvable via public `POST /invites/validate`), so that's real attack surface — storage-fill, enumeration — for data that isn't sensitive.
- It also had an **IP + coarse-UA fingerprint fallback** for iOS. Two users behind the same CGNAT or corporate egress on the same OS version can collide, and the loser gets someone else's invite code and destination. It's also the fingerprinting-for-attribution pattern Apple disallows, and Private Relay defeats it anyway. Not going there was right.
- No network call on first launch is a genuine robustness win — a slow API can't delay the cold start.

Three findings, one of them shipped as a PR:

**1. Telemetry — #2587 (stacked on this branch).** The restore path emits nothing today, and the clipboard `catch {}` swallows the one distinction that matters: a declined paste prompt and an organic install both end as `raw === null`. So a hand-off that never works on iOS is indistinguishable from one nobody used — the only signal either way is an absence of restores, which looks identical to success. That PR splits the empty cases into `no_handoff` / `marker_missing` / `clipboard_unavailable` and reports them with the channel. Only booleans leave the device (no invite code, no dest). It's stacked, so retarget it to `main` once this merges — happy to close it if you'd rather fold the change in here yourself.

**2. `/invite` dead-ends on iOS today.** Unrelated to your PR, but it hits the same path and it's the main deferred-linking use case, so worth knowing: the AASA is `paths: ["*"]` for all three appIDs, so `peanut.me/invite?code=…` routes into the app — but `/invite` is stripped from the native export (`scripts/native-build.js` ITEMS_TO_DISABLE) and isn't in the Android intent filter. On iOS with the app installed, an invite link lands on a route that doesn't exist.

The fix looks small: rewrite `/invite?code=X` → `/setup?step=signup` in the native mapper and write the code with the same normalization `applyDeferredPayload` already uses. Notably **no auth branch is needed** — `/setup` already detects arriving with an existing session and prompts continue/start-fresh, so a signed-in user tapping a shared invite link is handled. I haven't built it: it's a routing behaviour change on your ticket and would be a third stacked PR, so it seemed better as your call than my patch.

**3. Claim links — a warning about the obvious next step, not about this PR.** This PR sensibly carries only non-secret context. When claim links come up, please **don't** do what my plan proposed: it put `pnut1.<id>.<key>` — record id *and* AES key — into the referrer/clipboard, with ciphertext server-side. I'd convinced myself that kept the secret away from the server, Google and network observers, which is true, but it hands **both halves to anything that gets a pasteboard read on iOS**, and since the record is single-use the thief drains the link and the real user gets nothing. That's fund loss rather than a privacy nit. iOS 16's paste prompt mitigates but doesn't remove it. The existing "paste your link" recovery is a much safer fallback for the secret-bearing case.

Separately, #2584 fixes a **pre-existing** bug I hit while reading this: `mapDeepLink()` never reads `parsed.hash`, so it drops the fragment — meaning a claim link's `#p=<password>` is lost even on a *warm* deep link today, independent of deferred linking. Based on `dev`, doesn't block this PR, but it's a prerequisite before claim links could ever ride this path.

Nice work on the double-gated clipboard read and the one-shot flag ordering — the "set the flag before the clear so an interrupted clear can't re-read" detail is the kind of thing that bites later.

**Comment by kushagrasarathe** [comment]
Thanks @jjramirezn — all 10 triaged; 7 fixed in ffc9e79e8, 1 refuted with evidence, 2 accepted/declined with reasoning below.

**Fixed (1–5, 8, 10):**

1. **Login lockout** — restored `inviteCode` is now a **session cookie, byte-matching the web invite flow** (`InvitesPage` writes session too). Attribution survives the install→open→signup funnel (the case this feature exists for); an existing user who wants to log in self-heals on app restart instead of being stuck 30 days. `campaignTag` stays 30-day — it no longer gates the setup step (that was your #2346 fix), so it's safe to keep for badge attribution.
2. **Unmapped dest** — now must-map like `openDeepLink`: `deepLinkToNativePath` returning null (off-host, stray `%`) drops the dest instead of pushing it verbatim. Test added with a `/send/50%` dest.
3. **Transient referrer failure** — Android no longer consumes on a `null` read; the read is prompt-free and the referrer stays available ~90 days, so the next launch retries. A definitive read — including Play's organic `utm_source` string — consumes. Test covers the fail-then-succeed-next-launch sequence.
4. **Late-restore teleport** — dest navigation now only fires within **10s of launch** (`restoreStartedAt` window). A paste prompt answered minutes later, or a sluggish referrer service, applies cookies/locale but never yanks the user mid-flow. Hook test simulates a 20s-late resolve.
5. **Kill-during-prompt re-prompt** — the consumed flag is now set **before** `Clipboard.read()` (after the prompt-free gates pass). Killing the app mid-prompt loses the hand-off (lesser harm) but can never re-prompt. Test added.
8. **`/pt-br?x=1`** — `stripLocalePrefix` splits query/hash before segmenting. Test added.
10. **`toInviteCode`** — now called instead of duplicated.

**Refuted (6 — camera in the binary):** `scripts/native-release.sh:44` runs `pnpm exec cap sync android` on every release build, and `@capacitor/camera` has been in main's `package.json` since the QR-scanner feature — so **every shipped binary (v1.0.39–42) already contains the camera module**; the committed gradle files were just stale copies of what CI regenerates anyway. This PR changes nothing about the shipped manifest/permissions; it only brings the committed files in sync with what `cap sync` produces. (Play's sensitive-permission state is whatever v1.0.42 already declared.)

**Accepted, no code change (7 — invite-gate race):** when the referrer read lands after setup's ~100ms cookie check, the user sees Landing instead of the signup shortcut — but the cookie lands well before they finish tapping through, so invite/campaign attribution and the badge award are intact; only the gate-skip nicety degrades, on slow reads only. Wiring a re-check event into the setup page isn't worth the coupling for v1; happy to revisit if funnel data shows it matters.

**Declined (9 — dev-page wall fork):** the page must be reachable **unauthenticated in the native prod build** — it's the landing target for the fresh-install e2e (`dest=/dev/deferred`) — which `(mobile-ui)/dev` can't offer (auth layout), and generalizing `DevLayout` with a native-allowed flag for exactly one consumer is speculative abstraction. The 3-line wall names `(mobile-ui)/dev/layout.tsx` in its comment, so a future policy change greps straight to it.

**Comment by kushagrasarathe** [comment]
@innolope-dev thanks — this is a genuinely useful review, and the serverless-vs-server post-mortem is worth keeping (the CGNAT-collision + Private Relay points against fingerprinting deserve to be quotable next time someone proposes it). All three verified; here's where each lands:

**1. Telemetry (#2587)** — keep the PR, don't close it. [Reviewed here](https://github.com/peanutprotocol/peanut-ui/pull/2587#pullrequestreview-4815599598): direction and privacy shape are right, but it's stacked on `5231ab7d1` and today's review fixes (`ffc9e79e8`) rewrote the exact consumption block you instrument — Android now retries on a transient null read instead of consuming, so post-rebase your `no_handoff` would fire once per retry launch and skew the very match-rate it measures. Two asks in the review: rebase onto the current branch tip, and give the transient path its own `read_failed` outcome. Then retarget to `main` once #2560 merges.

**2. `/invite` dead-end — confirmed real.** `scripts/native-build.js:31` strips the `invite` dir from the export and the Android intent filter has no `/invite` entry, while the AASA is `paths: ["*"]` — so an iOS invite tap with the app installed lands nowhere. Taking your mapper-rewrite fix as a **follow-up PR on TASK-20772** (it's a routing behavior change; this PR is deep in review). One implementation note for whoever picks it up: the cookie write can't live in `mapDeepLink` — it runs during render in the notifications list (the URIError guard comment exists for that reason) — so the side effect belongs in `openDeepLink`, or the code rides a query param the setup page reads. Good catch that no auth branch is needed.

**3. Claim-link warning — agreed, and adopting it as a design constraint:** the deferred channel carries only non-secret, replayable context (locale / invite / campaign / dest). Anything secret-bearing — `#p=` passwords above all — stays on "re-tap the link" / paste-your-link recovery, where the iOS pasteboard-theft → drained-single-use-link failure mode you describe can't happen. Your own plan flagging its own flaw is the most useful kind of review.

Also read **#2584** — real pre-existing bug, and the fix shape is right (re-attach the fragment once above the branches so a new route can't forget it; the off-host reject still short-circuits before it). It composes cleanly with this PR: `sanitizeRedirectURL` already preserves hashes, and we don't touch `native-routes.ts` here, so no conflict in either merge order.

**Review by jjramirezn — APPROVED** [review]
